sockeye/model.py

```diff
@@ -154,26 +166,65 @@ def save_version(folder: str):
+                w_embed_source = w_embed_target = mx.sym.Variable(C.SHARED_EMBEDDING_PREFIX + "weight",
+                                                                  shape=(self.config.config_embed_source.vocab_size,
+                                                                         self.config.config_embed_source.num_embed))
+
+            if C.WEIGHT_TYING_SOFTMAX in self.config.weight_tying_type:
+                logger.info("Tying the target embeddings and output layer parameters.")
+                w_out_target = w_embed_target
+
+        return w_embed_source, w_embed_target, w_out_target
+
     def _build_model_components(self):
         """
         Instantiates model components.
         """
-        # we tie the source and target embeddings if both appear in the type
-        if self.config.weight_tying and C.WEIGHT_TYING_SRC in self.config.weight_tying_type \
-                and C.WEIGHT_TYING_TRG in self.config.weight_tying_type:
-            logger.info("Tying the source and target embeddings.")
-            embed_weight = encoder.Embedding.get_embed_weight(vocab_size=self.config.vocab_source_size,
-                                                              embed_size=0,  # will get inferred
-                                                              prefix=C.SHARED_EMBEDDING_PREFIX)
-        else:
-            embed_weight = None
-
-        self.encoder = encoder.get_encoder(self.config.config_encoder, embed_weight)
-
-        self.lexicon = lexicon.Lexicon(self.config.vocab_source_size,
-                                       self.config.vocab_target_size,
-                                       self.config.learn_lexical_bias) if self.config.lexical_bias else None
-
-        self.decoder = decoder.get_decoder(self.config.config_decoder, self.lexicon, embed_weight)
-
-        self.built = True
+        # encoder & decoder first (to know the decoder depth)
+        self.encoder = encoder.get_encoder(self.config.config_encoder)
+        self.decoder = decoder.get_decoder(self.config.config_decoder)
+
+        # source & target embeddings
+        embed_weight_source, embed_weight_target, out_weight_target = self._get_embed_weights()
+        assert isinstance(self.config.config_embed_source, encoder.EmbeddingConfig)
+        assert isinstance(self.config.config_embed_target, encoder.EmbeddingConfig)
+        self.embedding_source = encoder.Embedding(self.config.config_embed_source,
+                                                  prefix=C.SOURCE_EMBEDDING_PREFIX,
+                                                  embed_weight=embed_weight_source)
+        self.embedding_target = encoder.Embedding(self.config.config_embed_target,
+                                                  prefix=C.TARGET_EMBEDDING_PREFIX,
+                                                  embed_weight=embed_weight_target)
+
+        if self.config.weight_tying and C.WEIGHT_TYING_SOFTMAX in self.config.weight_tying_type:
```

Comment:
wouldn't it make more sense to have this check in `_get_embed_weights `?

Reply:
yes, will move it there